### PR TITLE
Implement ControlledGate._phase_by_

### LINKED ISSUE #1013: Add `_phase_by_` magic method to `ControlledGate`
Comes after https://github.com/quantumlib/Cirq/issues/947

The logic is as follows: if the qubit index is 0 (the control), the operation is returned unchanged. If it is larger then we delegate to phasing the sub gate with `cirq.phase_by` and a default result of NotImplemented. If it's NotImplemented, we return NotImplemented. Otherwise we return a controlled gate with the phased sub gate.

### PR DESCRIPTION
**Pull request by smitsanghavi**
Fixes #1013

### CODE REVIEW

**Comment by smitsanghavi** [comment]
Not sure how to test for the NotImplemented case. What would be a simple gate that doesn't SupportsPhase? Should I just create a dummy one?

**Comment by Strilanc** [comment]
Create a dummy one. There are existing gates that don't support this, such as the H gate, but we might add this functionality to any one of them at some point.

File: cirq/ops/controlled_gate_test.py [hunk 1]
```diff
@@ -219,6 +219,11 @@ def test_reversible():
     assert (cirq.inverse(cirq.ControlledGate(cirq.S)) ==
             cirq.ControlledGate(cirq.S**-1))
 
+def test_phase_by():
```
Comment (Strilanc) [inline cirq/ops/controlled_gate_test.py]: Add a test case where the underlying gate changes (e.g. use `cirq.google.ExpW`).

Comment (smitsanghavi) [reply cirq/ops/controlled_gate_test.py]: Done

File: cirq/ops/controlled_gate.py [hunk 2]
```diff
@@ -102,6 +96,15 @@ def __pow__(self, exponent: Any) -> 'ControlledGate':
             return NotImplemented
         return ControlledGate(new_sub_gate)
 
+    def _phase_by_(self, phase_turns: float, qubit_index: int):
+        if qubit_index == 0:
+            return self
+        phased_gate = protocols.phase_by(
+            self.sub_gate, phase_turns, qubit_index, None)
```
Comment (smitsanghavi) [inline cirq/ops/controlled_gate.py]: Should this be "qubit_index-1"?

Comment (Strilanc) [reply cirq/ops/controlled_gate.py]: Correct

Comment (Strilanc) [reply cirq/ops/controlled_gate.py]: Make sure to include a test that exercises this case, e.g. a Controlled(Controlled(NotPhaseable)).

**Comment by smitsanghavi** [comment]
Closing PR and reopening to retrigger the build checks that seem to be stuck

File: cirq/ops/controlled_gate_test.py [hunk 3]
```diff
@@ -176,6 +176,25 @@ def test_reversible():
...
+    sub_gate = cirq.google.ExpWGate(axis_half_turns = 0.5)
+    assert cirq.phase_by(sub_gate, 0.25, 1) != sub_gate
+    assert (cirq.phase_by(cirq.ControlledGate(sub_gate), 0.25, 0) ==
+            cirq.ControlledGate(sub_gate))
+    assert (cirq.phase_by(cirq.ControlledGate(sub_gate), 0.25, 1) !=
+            cirq.ControlledGate(sub_gate))
+    assert (cirq.phase_by(cirq.ControlledGate(sub_gate), 0.25, 1) ==
+            cirq.ControlledGate(cirq.phase_by(sub_gate, 0.25, 1)))
+    # Test that the qubit_index arg gets decremented at each subgate step.
+    assert (cirq.phase_by(
```
Comment (Strilanc) [inline cirq/ops/controlled_gate_test.py]: Check all three indices.

File: cirq/ops/controlled_gate_test.py [hunk 4]
```diff
@@ -176,6 +176,25 @@ def test_reversible():
...
+class UnphasableGate(cirq.SingleQubitGate):
+    pass
+
+def test_phase_by():
+    assert (cirq.phase_by(
+                cirq.ControlledGate(UnphasableGate), 0.25, 1, default=None) ==
+            None)
+    sub_gate = cirq.google.ExpWGate(axis_half_turns = 0.5)
+    assert cirq.phase_by(sub_gate, 0.25, 1) != sub_gate
+    assert (cirq.phase_by(cirq.ControlledGate(sub_gate), 0.25, 0) ==
```
Comment (Strilanc) [inline cirq/ops/controlled_gate_test.py]: Add a line `cg = cirq.ControlledGate(sub_gate)` to shorten these up.

**Pull request merged**